### DISCUSSION TITLE
rangeStrategy "replace" with cargo forces ~major.minor.patch for tilde requirements

### DISCUSSION THREAD

**Discussion by florent-gama** [renovatebot/renovate#27094] [Request Help] [answered]
### How are you running Renovate?

Self-hosted Renovate

### If you're self-hosting Renovate, tell us what version of Renovate you run.

37.168.1

### If you're self-hosting Renovate, select which platform you are using.

GitLab self-hosted

### Was this something which used to work for you, and then stopped?

I have not seen this working

### Describe the problem

When using `rangeStrategy: "replace"` with cargo dependencies specified with partial tilde requirements (eg. `"~0.1"` or `"~1"`), renovate will replace them with an exhaustive `"~major.minor.patch"` specification when an out-of-range update is available (eg. `"~0.2.0"` or `"~2.0.0"`), while I would have expected it to keep the specification format and bump the appropriate version fields (eg. `"~0.2"` or `"~2"`).

Minimal reproduction: https://github.com/florent-gama/renovate-reproduction-cargo-replace-partial-tilde

### Relevant debug logs

<details><summary>Logs</summary>

```

DEBUG: packageFiles with updates
{
  "baseBranch": "main"
  "config": {
    "cargo": [
      {
        "deps": [
          {
            "currentValue": "~0.7",
            "currentVersion": "0.7.3",
            "datasource": "crate",
            "depName": "rand",
            "depType": "dependencies",
            "dependencyUrl": "https://crates.io/crates/rand",
            "homepage": "https://rust-random.github.io/book",
            "isSingleVersion": false,
            "managerData": {
              "nestedVersion": false
            },
            "packageName": "rand",
            "registryUrl": "https://crates.io",
            "sourceUrl": "https://github.com/rust-random/rand",
            "versioning": "cargo",
            "warnings": [],
            "updates": [
              {
                "bucket": "non-major",
                "newVersion": "0.8.5",
                "newValue": "~0.8.0",
                "newMajor": 0,
                "newMinor": 8,
                "updateType": "minor",
                "isRange": true,
                "branchName": "renovate/rand-0.x"
              }
            ]
          }
        ],
        "packageFile": "Cargo.toml",
        "packageFileVersion": "0.1.0"
      }
    ]
  }
}

```

</details>

### Have you created a minimal reproduction repository?

I have linked to a minimal reproduction in the description above

**Comment by rarkins** [comment, marked as answer]
Fixed by https://github.com/renovatebot/renovate/pull/27386

**Reply by florent-gama** [reply]
Thanks for the fix!

I did not think about it while reporting the issue initially, but there is a similar issue with the "update-lockfile" strategy. I have created another minimal reproduction for that: https://github.com/florent-gama/renovate-reproduction-cargo-update-lockfile-partial-tilde
